addons/udes_stock/models/stock_move_line.py

```diff
@@ -0,0 +1,183 @@
+# -*- coding: utf-8 -*-
+from odoo import models, fields, api, _
+from collections import defaultdict
+from odoo.tools.float_utils import float_compare, float_round
+from odoo.osv import expression
+from odoo.exceptions import ValidationError
+
+
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    u_picking_type_id = fields.Many2one(
+        "stock.picking.type",
+        "Operation Type",
+        related="move_id.picking_type_id",
+        store=True,
+        readonly=True,
+    )
+
+    def get_lines_incomplete(self):
+        """Return the move lines in self that are not completed,
+        i.e., quantity done < quantity to do
+        """
+        return self.filtered(lambda ml: ml.qty_done < ml.product_uom_qty)
+
+    def get_lines_done(self):
+        """Return the move lines in self that are completed,
+        i.e., quantity done >= quantity to do
+        """
+        return self.filtered(lambda ml: ml.qty_done >= ml.product_uom_qty)
+
+    def move_lines_for_qty(self, quantity, sort=True):
+        """Return a subset of move lines from self where their sum of quantity
+        to do is equal to parameter quantity.
+        In case that a move line needs to be split, the new move line is
+        also returned (this happens when total quantity in the move lines is
+        greater than quantity parameter).
+        If there is not enough quantity to do in the move lines,
+        also return the remaining quantity.
+        """
+        new_ml = None
+        result = self.browse()
+        if quantity == 0:
```

Comment:
I can't remember - what would the use case of quantity==0 be here? Is it just due to iteration over something possibly?

Reply:
I have no idea

Reply:
no worries - i guess must be iteration. Else why would you call that function to return an empty move line :)